plugin.js

```diff
@@ -0,0 +1,112 @@
+'use strict';
+
+const PLUGIN_ID_KEY = '___shelljs_plugin_id___';
+const DEFAULT_ERROR_CODE = 1;
+
+function setPluginID(plugin, id) {
+  Object.defineProperty(plugin, PLUGIN_ID_KEY, {
+    value: id
+  });
+  return id;
+}
+
+const getPluginID = plugin => plugin[PLUGIN_ID_KEY];
+
+module.exports = function setupShellPlugins($) {
+  const used_plugins = [];
+  const plugins = [{ // $
+    id: 0,
+    name: '',
+    fullname: '',
+    func: $,
+    target: null
+  }];
+  setPluginID($, 0);
+
+  const state = {
+    error: null,
+    errorCode: 0,
+    currentCmd: 'shell.js',
+    tempDir: null,
+  };
+  $.config = {
+    silent: false,
+    fatal: false,
+    verbose: false,
+    noglob: false,
+    globOptions: false,
+    maxdepth: false,
+  };
+
+
+  const utils = {
+    platform: os.type().match(/^Win/) ? 'win' : 'unix',
+    plugin(target, fullname, func) {
```

Comment:
This function is actually what creates a new shelljs command. It would be invoked like this:

```javascript
$.use(($, utils) => {
  var test = utils.plugin($, 'test', msg => console.log('$.test: ' + msg));
  utils.plugin(test, 'testing', msg => console.log('$.test.testing: ' + msg));
  utils.plugin($.test, 'testing2', msg => console.log('$.test.testing2:' + msg));
  $.test === test; // true
});

// You can also extend existing commands.

$.use(($, utils) => {
    utils.plugin($.test, 'test.testing3', msg => console.log('$.test.testing3'));
});

$.test('foo!'); // $.test: foo!
$.test.testing('bar!'); // $.test.testing: bar!
$.test.testing2('baz!'); // $.test.testing2: baz!
$.test.testing3('quz!'); // $.test.testing3: quz!
```

There's currently no way to create commands that can't be invoked (ex. you can't just call `$.git()`). I'm thinking the syntax might be `utils.plugin($, 'whatever');`